core/node/state_keeper/src/executor/tests/tester.rs

```diff
@@ -149,16 +149,21 @@ impl Tester {
         l1_batch_env: L1BatchEnv,
         system_env: SystemEnv,
     ) -> Box<dyn BatchExecutor<OwnedStorage>> {
-        let mut batch_executor = MainBatchExecutorFactory::new(self.config.save_call_traces, false);
-        batch_executor.set_fast_vm_mode(self.config.fast_vm_mode);
-
         let (_stop_sender, stop_receiver) = watch::channel(false);
         let storage = storage_factory
             .access_storage(&stop_receiver, l1_batch_env.number - 1)
             .await
             .expect("failed creating VM storage")
             .unwrap();
-        batch_executor.init_batch(storage, l1_batch_env, system_env)
+        if self.config.trace_calls {
+            let mut executor = MainBatchExecutorFactory::<TraceCalls>::new(false);
+            executor.set_fast_vm_mode(self.config.fast_vm_mode);
+            executor.init_batch(storage, l1_batch_env, system_env)
+        } else {
+            let mut executor = MainBatchExecutorFactory::<()>::new(false);
+            executor.set_fast_vm_mode(self.config.fast_vm_mode);
+            executor.init_batch(storage, l1_batch_env, system_env)
```

Comment:
Put the if only around MainBatchExecutorFactory::new

Reply:
Do you mean `let mut executor = if ...`? It won't work like that because executors differ by the generic type (`<TracerCalls>` vs `<()>`).

Reply:
That is correct. The duplication can still be removed but the result might be ugly.